### PR TITLE
Bug 1853726: Out-of-tree driver with SRO implementation.

### PR DESCRIPTION
**Pull request by zvonkok**

### CODE REVIEW

**Comment by ArangoGutierrez** [comment]
testing this on a 4.5 cluster, not working, even after adding entitlements

**Comment by ArangoGutierrez** [comment]
Pinging this here - https://www.openshift.com/blog/how-to-use-entitled-image-builds-to-build-drivercontainers-with-ubi-on-openshift

File: pkg/controller/specialresource/controller.go [hunk 1]
```diff
@@ -176,12 +179,40 @@ func (r *ReconcileSpecialResource) Reconcile(request reconcile.Request) (reconci
...
 	}
 
-	if err := ReconcileHardwareConfigurations(r); err != nil {
-		// We do not want a stacktrace here, errs.Wrap already created
-		// breadcrumb of errors to follow. Just sprintf with %v rather than %+v
-		log.Info("Could not reconcile hardware configurations", "error", fmt.Sprintf("%v", err))
-		return reconcile.Result{}, errs.New("Reconciling failed")
+	for _, specialresource := range specialresources.Items {
+
+		log.Info("Reconciling", "SpecialResurce", specialresource.Name)
```
Comment (kpouget) [inline pkg/controller/specialresource/controller.go]: typo here and line below: `s/SpecialResurce/SpecialResource/`

**Comment by kpouget** [comment]
I find it quite hard to review such a big patch; it seems that you squashed multiple patches together, maybe you could include each of the in the PR?

**Comment by kpouget** [comment]
I find it quite hard to review such a big patch; it seems that you squashed multiple patches together, maybe you could include each of the in the PR?

**Comment by kpouget** [comment]
Apart from the details mentioned in the comments, the patch worked well for me, the operator managed to compile the driver after installing the kernel files from the payload RPMs.

```
Labels:
                    beta.kubernetes.io/instance-type=p3.2xlarge
                    feature.node.kubernetes.io/kernel-version.full=4.18.0-147.8.1.el8_1.x86_64
                    feature.node.kubernetes.io/pci-10de.present=true
                    specialresource.openshift.io/device-monitoring-nvidia-gpu=ready
                    specialresource.openshift.io/device-plugin-nvidia-gpu=ready
                    specialresource.openshift.io/driver-container-nvidia-gpu=ready
                    specialresource.openshift.io/runtime-enablement-nvidia-gpu=ready
                    [...]
Capacity:
  [...]
  nvidia.com/gpu:              1
```
(note that is was already working before on this cluster version (`OC 4.4.8`))

File: pkg/controller/specialresource/controller.go (hunk 1 above)
Comment (zvonkok) [reply pkg/controller/specialresource/controller.go]:
```suggestion
		log.Info("Reconciling", "SpecialResource", specialresource.Name)
```

**Comment by ArangoGutierrez** [comment]
`test images` is failing do Conflict in merge 
```bash
{"command":"git merge --no-ff 71c884d2469e9ccaf0be210bc9fb8a51db68548a","component":"clonerefs","error":"exit status 1","file":"prow/pod-utils/clone/clone.go:61","func":"k8s.io/test-infra/prow/pod-utils/clone.Run.func2","level":"info","msg":"Ran command","output":"Auto-merging vendor/modules.txt\nAuto-merging vendor/github.com/openshift/api/security/v1/zz_generated.swagger_doc_generated.go\nAuto-merging vendor/github.com/openshift/api/route/v1/zz_generated.swagger_doc_generated.go\nAuto-merging vendor/github.com/openshift/api/image/v1/zz_generated.swagger_doc_generated.go\nAuto-merging vendor/github.com/openshift/api/build/v1/zz_generated.swagger_doc_generated.go\nRemoving recipes/nvidia-vran/config/mofed-patch.json\nRemoving recipes/nvidia-vran/config/gpudirect-patch.json\nRemoving recipes/nvidia-vran/config/gdrdrv-patch.json\nRemoving recipes/nvidia-gpu/config/patch.json\nAuto-merging recipes/mlnx-rdma-shared/manifests/mofed/1000-state-driver.yaml\nAuto-merging recipes/mlnx-rdma-shared/manifests/gpudirect/1000-state-driver.yaml\nRemoving recipes/mellanox-rdma-sriov/manifests/mofed/1000-state-driver.yaml\nRemoving recipes/mellanox-rdma-sriov/manifests/mofed/0000-state-driver-buildconfig.yaml\nRemoving recipes/mellanox-rdma-sriov/manifests/gpudirect/1000-state-driver.yaml\nRemoving recipes/mellanox-rdma-sriov/config/mofed-patch.json\nRemoving recipes/mellanox-rdma-sriov/config/gpudirect-patch.json\nRemoving recipes/mellanox-rdma-sriov/config/Makefile\nRemoving recipes/mellanox-rdma-shared/manifests/mofed/0000-state-driver-buildconfig.yaml\nRemoving recipes/mellanox-rdma-shared/config/mofed-patch.json\nRemoving recipes/mellanox-rdma-shared/config/gpudirect-patch.json\nRemoving recipes/mellanox-rdma-shared/config/Makefile\nAuto-merging go.sum\nCONFLICT (content): Merge conflict in go.sum\nAuto-merging go.mod\nCONFLICT (content): Merge conflict in go.mod\nRemoving deploy/crds/sro_v1alpha1_specialresource_crd.yaml\nRemoving deploy/crds/sro_v1alpha1_specialresource_cr.yaml\nAutomatic merge failed; fix conflicts and then commit the result.\n","severity":"info","time":"2020-07-21T15:33:17Z"}
```

**Review by ArangoGutierrez — APPROVED** [review]
/lgtm

**Pull request merged**